pkg/reconciler/pipelinerun/metrics.go

```diff
@@ -0,0 +1,157 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"knative.dev/pkg/metrics"
+)
+
+var (
+	prDuration = stats.Float64(
+		"pipelinerun_duration_seconds",
+		"The pipelinerun execution time in seconds",
+		stats.UnitDimensionless)
+	pRunDistributions = view.Distribution(10, 30, 60, 300, 900, 1800, 3600, 5400, 10800, 21600, 43200, 86400)
+
+	prCount = stats.Float64("pipelinerun_count",
+		"number of pipelineruns",
+		stats.UnitDimensionless)
+
+	runningPrsCount = stats.Float64("running_pipelineruns_count",
+		"Number of pipelineruns executing currently",
+		stats.UnitDimensionless)
+)
+
+type Recorder struct {
+	initialized bool
+
+	pipeline    tag.Key
+	pipelineRun tag.Key
+	namespace   tag.Key
+	status      tag.Key
+}
+
+func NewRecorder() (*Recorder, error) {
+	r := &Recorder{
+		initialized: true,
+	}
+
+	pipeline, err := tag.NewKey("pipeline")
+	if err != nil {
+		return nil, err
+	}
+	r.pipeline = pipeline
+
+	pipelineRun, err := tag.NewKey("pipelinerun")
+	if err != nil {
+		return nil, err
+	}
+	r.pipelineRun = pipelineRun
+
+	namespace, err := tag.NewKey("namespace")
+	if err != nil {
+		return nil, err
+	}
+	r.namespace = namespace
+
+	status, err := tag.NewKey("status")
+	if err != nil {
+		return nil, err
+	}
+	r.status = status
+
+	err = view.Register(
+		&view.View{
+			Description: prDuration.Description(),
+			Measure:     prDuration,
+			Aggregation: pRunDistributions,
+			TagKeys:     []tag.Key{r.pipeline, r.pipelineRun, r.namespace, r.status},
+		},
+		&view.View{
+			Description: prCount.Description(),
+			Measure:     prCount,
+			Aggregation: view.Count(),
+			TagKeys:     []tag.Key{r.status},
+		},
+		&view.View{
+			Description: runningPrsCount.Description(),
+			Measure:     runningPrsCount,
+			Aggregation: view.LastValue(),
+		},
+	)
+
+	if err != nil {
+		r.initialized = false
+		return r, err
+	}
+
+	return r, nil
+}
+
+func (r *Recorder) DurationAndCount(pr *v1alpha1.PipelineRun) error {
+	if !r.initialized {
+		return fmt.Errorf("ignoring the metrics recording for %s , failed to initialize the metrics recorder", pr.Name)
+	}
+
+	duration := time.Since(pr.Status.StartTime.Time)
+	if pr.Status.CompletionTime != nil {
+		duration = pr.Status.CompletionTime.Sub(pr.Status.StartTime.Time)
+	}
+
+	status := "success"
+	if pr.Status.Conditions[0].Status == corev1.ConditionFalse {
+		status = "failed"
+	}
+
+	ctx, err := tag.New(
+		context.Background(),
+		tag.Insert(r.pipeline, pr.Spec.PipelineRef.Name),
+		tag.Insert(r.pipelineRun, pr.Name),
+		tag.Insert(r.namespace, pr.Namespace),
+		tag.Insert(r.status, status),
+	)
+
+	if err != nil {
+		return err
+	}
+
+	metrics.Record(ctx, prDuration.M(float64(duration/time.Second)))
+	metrics.Record(ctx, prCount.M(1))
+
+	return nil
+}
+
+func (r *Recorder) RunningPrsCount(lister listers.PipelineRunLister) error {
+	if !r.initialized {
+		return errors.New("ignoring the metrics recording, failed to initialize the metrics recorder")
+	}
+
+	prs, err := lister.List(labels.Everything())
```

Comment:
This might take a _very_ long time if there are many many PipelineRuns in the history. Can we filter for complete ones, or sort by completion time and stop when we see a completed run?

Reply:
Yes @ImJasonH I was also concerned about it while using in the first place. However its retrieving data from the in-memory cache. 
Lets try to sort the data and check if its optimze the time.

Safer option would be to run this asynchrousnly (in another goroutine)?

Reply:
Let's run this in another goroutine, so that this doesn't contribute to reconciliation slowness. I'm glad it's coming from an in-memory cache, but this might still take long enough to matter, especially if some things aren't guaranteed to be in the cache.

Reply:
Yes now we are logging metrics asynchronously. :+1: